server/src/main/java/org/elasticsearch/rest/RestRequest.java

```diff
@@ -306,9 +303,27 @@ public boolean isFullContent() {
         return httpRequest.body().isFull();
     }
 
+    /**
+     * Returns a copy of HTTP content. The copy is GC-managed and does not require reference counting.
+     * Please use {@link #releasableContent()} to avoid content copy.
+     */
     public BytesReference content() {
+        return BytesReference.copyBytes(releasableContent());
+    }
+
+    /**
+     * Returns a reference to the network buffer of HTTP content.
+     */
+    public ReleasableBytesReference releasableContent() {
         this.contentConsumed = true;
-        return httpRequest.body().asFull().bytes();
+        var bytes = httpRequest.body().asFull().bytes();
+        assert bytes.hasReferences() : AbstractRefCounted.ALREADY_CLOSED_MESSAGE;
+        if (bytes.hasReferences() == false) {
+            var e = new IllegalStateException(AbstractRefCounted.ALREADY_CLOSED_MESSAGE);
+            logger.error(e.getMessage(), e);
+            throw e;
```

Comment:
Suggest moving the assertion here, after logging, so that we see the bug in `stdout` in tests:

```suggestion
            assert false : e;
            throw e;
```

Reply:
Done